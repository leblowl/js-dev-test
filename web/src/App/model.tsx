const ANY_LANGUAGE = 'All';

const initialModel = {
  entities: {
    repos: []
  },
  repoList: {
    languages: [],
    languageSelected: ANY_LANGUAGE,
    repos: []
  }
};

export { ANY_LANGUAGE, initialModel };
