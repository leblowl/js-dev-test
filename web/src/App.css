html, body {
  width: 100%;
  min-width: 100%;
  max-width: 100%;

  height: 100%;
  min-height: 100%;
  max-height: 100%;
}

ul, li, div {
  box-sizing: border-box;
}

button {
  border: none;
  background: transparent;
  cursor: pointer;
}

ul {
  list-style-type: none;
  background: #eee;
  margin: 0;
  padding: 0;
}

.button-row {
  padding: 1em;
}

.button-row .selected {
  border: 1px solid blue;
}

.repo-list {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 1em;
  border-radius: 4px;
}

.repo-item {
  padding: 1em;
  background: #fafafa;
  border: 1px solid #ccc;
  width: 100%;
  margin: 1em 1em;
  border-radius: 4px;
  display: flex;
  justify-content: flex-start;
}

.repo-item button.name {
  color: #0099FF;
  text-align: left;
  max-width: 20vw;
  flex: 1;
}

.repo-item span.description {
  flex: 1;
}

.repo-item span {
  margin: 0 1em;
  text-align: left;
}

.repo-item.selected {
  border: 1px solid blue;
}

.App {
  padding: 1em;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
  pointer-events: none;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
