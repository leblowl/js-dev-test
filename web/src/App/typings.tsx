export interface Author extends Object {
  name: string;
  date: string;
}

export interface Commit extends Object {
  author: Author;
  message: string;
}

export interface Repo extends Object {
  id: number;
  selected: boolean;
  name: string;
  full_name: string;
  description: string;
  language: string;
  forks_count: number;
  latestCommit: Commit
}

export interface Entities extends Object {
  repos: Array<Repo>;
}

export interface RepoListVM extends Object {
  languages: Array<string>;
  languageSelected: string;
  repos: Array<Repo>;
}

export interface AppState extends Object {
  entities: Entities;
  repoList: RepoListVM;
}

export interface EmitApp extends React.Component<Object, AppState> {
  emit: Function;
}
