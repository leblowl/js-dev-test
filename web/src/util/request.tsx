const get = function(url: string, onload: Function) {
  let req = new XMLHttpRequest();

  req.responseType = 'json';
  req.open('GET', url);
  req.onload = function(e) {
    if (this.status === 200) {
      onload(this.response);
    }
  };
  req.send();
}

export { get };

