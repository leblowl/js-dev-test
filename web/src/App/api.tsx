import { Repo } from './typings';
import * as request from '../util/request';

/**
 * URL for application backend repos.
 */
const reposUrl = 'http://localhost:4000/repos';

/**
 * Fetch repos from application backend.
 */
const getRepos = function(onLoad: Function, onErr: Function) {
  request.get(reposUrl, onLoad, onErr);
}

/**
 * Get Github latest commits URL for a given repo.
 */
const repoCommitUrl = function(repo: Repo) {
  return `https://api.github.com/repos/${repo.full_name}/commits`;
}

/**
 * Fetch latest repo commits from Github.
 */
const getLatestRepoCommits = function(repo: Repo, onLoad: Function, onErr: Function) {
  request.get(repoCommitUrl(repo), onLoad, onErr);
}

export { getRepos, getLatestRepoCommits }
