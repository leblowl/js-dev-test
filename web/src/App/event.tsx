import { EmitApp, Repo, Commit } from './typings';
import * as model from './model';
import * as request from '../util/request';


function onlyUnique(value: any, index: number, self: Array<any>) {
      return self.indexOf(value) === index;
}

const repoCommitUrl = function(repo: Repo) {
  return `https://api.github.com/repos/${repo.full_name}/commits`;
}

const onReceiveLatestCommits = function(
  app: EmitApp,
  data: any
) {
  app.setState((state, props) => {
    let ndx = state.entities.repos.indexOf(data.repo);

    state.entities.repos[ndx].latestCommit = (data.data[0] && data.data[0].commit) || {
      'author': {'name': 'test', 'date': 'now'},
      'message': 'This is a placeholder message for when Github unblocks me'
    };
    return state;
  });
}

const getLatestRepoCommit = function(repo: Repo, onLoad: Function, onErr: Function) {
  request.get(repoCommitUrl(repo), onLoad, onErr);
}

const onRequestLatestCommits = function(
  app: EmitApp,
  repos: Array<Repo>
) {
  repos.forEach((repo) => {
    let onResp = (data: Array<Commit>) => {
      app.emit(onReceiveLatestCommits, { repo, data });
    }
    getLatestRepoCommit(repo, onResp, onResp);
  });
}

const onInitState = function(
  app: EmitApp,
  data: Array<Repo>
) {
  app.setState((state, props) => {
    state.entities.repos = data;
    state.repoList.repos = state.entities.repos;

    let uniqueLanguages = [model.ANY_LANGUAGE].concat(state.entities.repos.map(
      (x) => x.language).filter(onlyUnique));
    state.repoList.languages = uniqueLanguages;

    return state;
  }, () => app.emit(onRequestLatestCommits, app.state.entities.repos));
}

const reposUrl = 'http://localhost:4000/repos';

const getRepos = function(onLoad: Function, onErr: Function) {
  request.get(reposUrl, onLoad, onErr);
}

const onAppMount = function(
  app: EmitApp,
  data: Array<Repo>
) {
  getRepos((data: Array<Repo>) => {
    app.emit(onInitState, data);
  }, console.error);
}

const onSelectRepo = function(
  app: EmitApp,
  id: number
) {
  app.setState((state, props) => {
    state.repoList.repos = state.entities.repos.map((x) => {
      if (x.id === id) {
        x.selected = !x.selected;
        return x;
      } else {
        return x;
      }});
    return state;
  });
}

const onSortByLanguage = function(
  app: EmitApp,
  language: string
) {
  app.setState((state, props) => {
    if (language === model.ANY_LANGUAGE) {
      state.repoList.repos = state.entities.repos;
    } else {
      state.repoList.repos = state.entities.repos.filter((x) => {
        return x.language === language;
      })
    }
    state.repoList.languageSelected = language;
    return state;
  });
}

export { reposUrl, getRepos, getLatestRepoCommit, onAppMount, onInitState, onSelectRepo, onSortByLanguage };
