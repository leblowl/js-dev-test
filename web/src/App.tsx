import React from 'react';
import './App.css';
import { AppState, Repo, RepoListVM } from './App/typings';
import * as model from './App/model';
import * as event from './App/event';


const RepoItemDetails: React.FunctionComponent<{
  repo: Repo;
  emit: Function;
}> = function({ repo, emit }) {
  if (repo.latestCommit) {
    return (
      <div className='body'>
        <span>{'AUTHOR: ' + repo.latestCommit.author.name}</span>
        <span>{'DATE: ' + repo.latestCommit.author.date}</span>
        <span>{'MESSAGE: ' + repo.latestCommit.message}</span>
      </div>
    );
  } else {
    return <div className='body'></div>;
  }
}

const RepoItem: React.FunctionComponent<{
  repo: Repo;
  emit: Function;
}> = function({ repo, emit }) {

  let selectedClass = repo.selected ? 'selected' : '';
  return (
    <li className={'repo-item ' + selectedClass}>
      <div className='header'>
        <button
          className='name'
          onClick={() => emit(event.onSelectRepo, repo.id)}>
          {repo.name}
        </button>
        <span className='description'>{repo.description}</span>
        <span>{repo.language}</span>
        <span>{repo.forks_count}</span>
      </div>
      <RepoItemDetails repo={repo} emit={emit}/>
    </li>
  );
};

const LanguageButtons: React.FunctionComponent<{
  languages: Array<string>;
  languageSelected: string;
  emit: Function;
}> = function({ languages, languageSelected, emit }) {

  return (
    <div className='button-row'>
      {languages.map((x) => {
        return (
          <button
            key={x}
            className={x === languageSelected ? 'selected' : ''}
            onClick={() => emit(event.onSortByLanguage, x)}>
            {x}
          </button>
        );
      })}
    </div>
  )
}

const RepoList: React.FunctionComponent<{
  repoList: RepoListVM;
  emit: Function;
}> = function( { repoList, emit }) {
  return (
    <div className='repo-list'>
      <LanguageButtons
        languages={repoList.languages}
        languageSelected={repoList.languageSelected}
        emit={emit} />

      <ul className='repo-list'>
        {repoList.repos.map((x) => {
          return <RepoItem key={x.id} repo={x} emit={emit} />
        })}
      </ul>
    </div>
  );
}

class App extends React.Component<Object, AppState> {
  constructor(props: Object) {
    super(props);
    this.state = model.initialModel;
  }

  emit(event: Function, data: Object) {
    event(this, data);
  }

  componentDidMount() {
    this.emit(event.onAppMount, {});
  }

  render() {
    let emit = (event: Function, data: Object) => this.emit(event, data);

    return (
      <div className='App'>
        <RepoList
          repoList={this.state.repoList}
          emit={emit} />
      </div>
    );
  }
}

export default App;
