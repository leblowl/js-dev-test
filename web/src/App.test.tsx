/* eslint-env jest */

import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import renderer from 'react-test-renderer';
import * as request from './util/request';
import * as event from './App/event.tsx';
const { act } = renderer;

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});

test('App displays repos', async () => {
  let comp;

  const fakeRepos = [
    {
     "id": 12345678,
     "name": "JSONExample",
     "full_name": "silverorange/JSON-Example",
     "description": "An example repository from a JSON file.",
     "language": "English",
     "forks_count": 999
    },
    {
     "id": 87654321,
     "name": "JSONExample2",
     "full_name": "silverorange/JSON-Example2",
     "description": "An SECOND example repository from a JSON file.",
     "language": "French",
     "forks_count": 888
    }
  ];

  const fakeLatestCommit = [
      {
        'commit': {
          'author': {'name': 'SuperWoman', 'date': 'Tomorrow'},
          'message': 'Huh?'
        }
      }
  ];

  // NOTE: Couldn't get spyOn to work for event.getRepo...
  // So I am simply spying on request.get for now... adding a little conditional
  // on the URL...
  // TODO: Fixme!
  jest.spyOn(request, 'get').mockImplementation((url, onLoad, onError) => {
    // This is pretty resilient
    if (url === event.reposUrl) {
      onLoad(fakeRepos);
    }
    // This is very fragile
    if (url.startsWith('https://api.github.com/repos/')) {
      onLoad(fakeLatestCommit);
    }
  });

  await act(async () => {
    comp = renderer.create(<App />);
  });

  let tree = comp.toJSON();
  expect(tree).toMatchSnapshot();
});
