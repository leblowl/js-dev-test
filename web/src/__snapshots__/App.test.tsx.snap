// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App displays repos 1`] = `
<div
  className="App"
>
  <div
    className="repo-list"
  >
    <div
      className="button-row"
    >
      <button
        className="selected"
        onClick={[Function]}
      >
        All
      </button>
      <button
        className=""
        onClick={[Function]}
      >
        English
      </button>
      <button
        className=""
        onClick={[Function]}
      >
        French
      </button>
    </div>
    <ul
      className="repo-list"
    >
      <li
        className="repo-item "
      >
        <div
          className="header"
        >
          <button
            className="name"
            onClick={[Function]}
          >
            JSONExample
          </button>
          <span
            className="description"
          >
            An example repository from a JSON file.
          </span>
          <span>
            English
          </span>
          <span>
            999
          </span>
        </div>
        <div
          className="body"
        >
          <span>
            AUTHOR: SuperWoman
          </span>
          <span>
            DATE: Tomorrow
          </span>
          <span>
            MESSAGE: Huh?
          </span>
        </div>
      </li>
      <li
        className="repo-item "
      >
        <div
          className="header"
        >
          <button
            className="name"
            onClick={[Function]}
          >
            JSONExample2
          </button>
          <span
            className="description"
          >
            An SECOND example repository from a JSON file.
          </span>
          <span>
            French
          </span>
          <span>
            888
          </span>
        </div>
        <div
          className="body"
        >
          <span>
            AUTHOR: SuperWoman
          </span>
          <span>
            DATE: Tomorrow
          </span>
          <span>
            MESSAGE: Huh?
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
